sycl/include/sycl/ext/intel/experimental/esimd/math.hpp

```diff
@@ -1667,6 +1667,23 @@ ESIMD_NODEBUG ESIMD_INLINE
   return esimd_pack_mask(src_0);
 }
 
+/// Compare source vector elements against zero and return a bitfield combining
+/// the comparison result. The representative bit in the result is set if
+/// corresponding source vector element is non-zero, and is unset otherwise.
+/// @param mask the source operand to be compared with zero.
+/// @return an \c uint, where each bit is set if the corresponding element of
+/// the source operand is non-zero and unset otherwise.
+template <typename T, int N>
+ESIMD_NODEBUG ESIMD_INLINE typename sycl::detail::enable_if_t<
+    detail::is_type<T, ushort, uint> && (N > 0 && N <= 32), uint>
+esimd_ballot(simd<T, N> mask) {
+  simd_mask<N> cmp = (mask != 0);
+  constexpr int N1 = (N <= 8 ? 8 : N <= 16 ? 16 : 32);
+  simd<uint16_t, N1> src0 = 0;
+  src0.template select<N, 1>() = cmp.data();
```

Comment:
After our discussion, it occurred to me that `src0.template select<N, 1>() = cmp.data()` might not be free and generate some extra unnecessary code if N is either 8, 16 or 32 - (zero initialization plus __esimd_vstore). I'd suggest to specialize this for 8, 16 or 32 cases and omit src0 manipulation there. Does this make sense?